src/wallet/wallet.cpp

```diff
@@ -3919,69 +3262,41 @@ std::set<CTxDestination> CWallet::GetLabelAddresses(const std::string& label) co
 
 bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestination& dest, bool internal)
 {
-    if (!pwallet->CanGetAddresses(internal)) {
+    m_spk_man = pwallet->GetScriptPubKeyMan(type, internal);
+    if (!m_spk_man) {
         return false;
     }
 
+    m_spk_man->TopUp();
```

Comment:
This TopUp seems newly added rather than moved, and the return value isn't checked. Tests also seem to pass without it. I also noticed TopUp calls were removed in `GetNewChangeDestination` and `CreateWalletFromFile`, and I don't know why.

These changes seem ok, but I guess the things I'd like to know would be:

1. If something in particular motivated these changes, or if they just seemed like a good ideas.
2. In general, if there's some kind of reasoning about where the keypool should be topped up. For example, I'd expect either keyman objects to be responible for topping up internally, or for the wallet to be responsible, but not for calls to happen both places. If there is intent behind this, maybe the Topup method could have a comment saying when it's supposed to be called.

Could also consider reverting top up changes that aren't necessary here. I think I'll be done with my neverending review of this PR soon, but little changes like this I don't understand do not make the end seem closer!  :smiley:

Reply:
This particular `TopUp` was moved from `GetNewChangeDestination` into `GetReservedDestination` because `GetNewChangeDestination` does not determine which `ScriptPubKeyMan` to use, but `GetReservedDestination` does.

Where in `CreateWalletFromFile` were `TopUp`s removed? It's likely that those were rolled into `SetupGeneration`.

Reply:
Ok, thanks for the explanation. The changes here do make sense since they're just moves. I didn't realize the GetNewChangeDestination and GetReservedDestination changes were related, or that SetupGeneration calls TopUp (indirectly through LegacyScriptPubKeyMan::NewKeyPool).